### PR TITLE
Add bash script to build each commit of a branch

### PR DESCRIPTION
**Pull request by ubruhin**
This script allows to automatically build each single commit of the current branch (up to the master branch).

### CODE REVIEW

File: dev/build_each_commit.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+START="master"
+END=${HEAD}
+
+trap "git checkout ${HEAD} && test ${STASH} -ne 1 || git stash pop" EXIT
+
+readonly COMMITS=$(git log --pretty=oneline "${START}...${END}" | cut -d' ' -f1 | tac)
+for commit in ${COMMITS[@]}; do
+	git checkout $commit
+	qmake -r ../librepcb.pro > "build_${commit}.log" 2>&1
+	make -j8 >> "build_${commit}.log" 2>&1
```
Comment (dns2utf8) [inline dev/build_each_commit.sh]: Switch to autodetect:
```
make -j$(nproc)
```

File: dev/build_each_commit.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (dns2utf8) [inline dev/build_each_commit.sh]: Use `#!/usr/bin/env bash`

**Comment by ubruhin** [comment]
> cool script, what is your usecase?

I often re-order commits in feature branches (i.e. `git rebase -i master`) to cleanup the history. But sometimes this introduces some errors, so that some commits won't compile anymore. With this script I can easily verify that each commit of the feature branch still compiles, so reordering didn't break something.

File: dev/build_each_commit.sh (hunk 2 above)
Comment (ubruhin) [inline dev/build_each_commit.sh]: Fixed.

File: dev/build_each_commit.sh (hunk 1 above)
Comment (ubruhin) [inline dev/build_each_commit.sh]: Fixed.

**Comment by dns2utf8** [comment]
🌈

File: dev/build_each_commit.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	STASH=1
+fi
+echo "STASH=${STASH}"
+
+START="master"
+END=${HEAD}
+
+trap "git checkout ${HEAD} && test ${STASH} -ne 1 || git stash pop" EXIT
+
+readonly COMMITS=$(git log --pretty=oneline "${START}...${END}" | cut -d' ' -f1 | tac)
```
Comment (dbrgn) [inline dev/build_each_commit.sh]: Whaat? I've never sean `readonly` :) That's pretty cool.

Comment (ubruhin) [inline dev/build_each_commit.sh]: I also didn't know that, I have stolen it from the Internet xD

**Comment by ubruhin** [comment]
> What happens if your branch is not based on master? Or if you have a detached head?

The command `git log --pretty=oneline "master...HEAD"` may fail in that case, but that's no problem because the `trap` command will always restore all changes made by this script. It doesn't matter whether the script was successful or not.

Because it's only a developer tool (basically for myself), I didn't add too much features and error handling ;)

**Comment by dbrgn** [comment]
Woah, I didn't know `trap` either. @ubruhin is the bash master :wink:

**Comment by rnestler** [comment]
Minor nit-pick: Why add this to LibrePCB if it is a personal developer tool? Or in other words: I would like to use this on other projects and have it directly in `$HOME/bin` :smiley:

**Comment by ubruhin** [comment]
> Why add this to LibrePCB if it is a personal developer tool? Or in other words: I would like to use this on other projects and have it directly in $HOME/bin

This script is explicitly designed to work with LibrePCB, the build steps will not work for other projects. So it makes more sense in this repository than in $HOME. And even if I created it for myself, this does not mean that nobody else can use it too ;)

Or should I change the whole repository into a private project because I'm the only user of it (or better delete it completely, because actually I also don't use it)? :smile:

**Pull request merged**